value.c

```diff
@@ -67,19 +74,25 @@ bool valtostr(const value_t * val, char *str, size_t n)
     else if (val->flags.s8b ) { max_bytes = 1; print_as_unsigned = false; }
 
     /* find the right format, considering different integer size implementations */
-         if (max_bytes == sizeof(long long)) snprintf(str, n, print_as_unsigned ? "%llu, %s" : "%lld, %s", print_as_unsigned ? get_ulonglong(val) : get_slonglong(val), buf);
-    else if (max_bytes == sizeof(long))      snprintf(str, n, print_as_unsigned ? "%lu, %s"  : "%ld, %s" , print_as_unsigned ? get_ulong(val) : get_slong(val), buf);
-    else if (max_bytes == sizeof(int))       snprintf(str, n, print_as_unsigned ? "%u, %s"   : "%d, %s"  , print_as_unsigned ? get_uint(val) : get_sint(val), buf);
-    else if (max_bytes == sizeof(short))     snprintf(str, n, print_as_unsigned ? "%hu, %s"  : "%hd, %s" , print_as_unsigned ? get_ushort(val) : get_sshort(val), buf);
-    else if (max_bytes == sizeof(char))      snprintf(str, n, print_as_unsigned ? "%hhu, %s" : "%hhd, %s", print_as_unsigned ? get_uchar(val) : get_schar(val), buf);
-    else if (val->flags.f64b) snprintf(str, n, "%lf, %s", get_f64b(val), buf);
-    else if (val->flags.f32b) snprintf(str, n, "%f, %s", get_f32b(val), buf);
-    else {
-        snprintf(str, n, "%#llx?, %s", get_slonglong(val), buf);
-        return false;
-    }
- 
-    return true;
+         if (max_bytes == sizeof(long long)) np = snprintf(str, n - np, print_as_unsigned ? "%llu" : "%lld", print_as_unsigned ? get_ulonglong(val) : get_slonglong(val));
+    else if (max_bytes == sizeof(long))      np = snprintf(str, n - np, print_as_unsigned ? "%lu"  : "%ld" , print_as_unsigned ? get_ulong(val) : get_slong(val));
+    else if (max_bytes == sizeof(int))       np = snprintf(str, n - np, print_as_unsigned ? "%u"   : "%d"  , print_as_unsigned ? get_uint(val) : get_sint(val));
+    else if (max_bytes == sizeof(short))     np = snprintf(str, n - np, print_as_unsigned ? "%hu"  : "%hd" , print_as_unsigned ? get_ushort(val) : get_sshort(val));
+    else if (max_bytes == sizeof(char))      np = snprintf(str, n - np, print_as_unsigned ? "%hhu" : "%hhd", print_as_unsigned ? get_uchar(val) : get_schar(val));
+    else if (val->flags.f64b) np = snprintf(str, n - np, "%lf", get_f64b(val));
+    else if (val->flags.f32b) np = snprintf(str, n - np, "%f", get_f32b(val));
+    else
+        goto err;
+    if (np <= 0 || np >= (n - bytes_in_buf))
+        goto err;
+    np = snprintf(str + np, bytes_in_buf, "%s", buf);
+    if (np <= 0)
+        goto err;
```

Comment:
Hmm, I think now this can be simplified by not splitting this off from the number printing and just checking if (np >= n) afterwards. In that case jump into the error case. I'll rework, re-test and forcefully push this.